src/circuit/dbsp_handle.rs

```diff
@@ -0,0 +1,309 @@
+use crate::{
+    circuit::runtime::RuntimeHandle, Circuit, Error as DBSPError, Runtime, RuntimeError,
+    SchedulerError,
+};
+use crossbeam::channel::{bounded, Receiver, Sender, TryRecvError};
+use std::thread::Result as ThreadResult;
+
+impl Runtime {
+    /// Instantiate and run a circuit in a multithreaded runtime.
+    ///
+    /// Creates a multithreaded runtime with `nworkers` worker threads,
+    /// instantiates identical circuits in each worker, using `constructor`
+    /// closure, and returns a [`DBSPHandle`] that can be used to control
+    /// the circuit.
+    pub fn run_circuit<F, T>(nworkers: usize, constructor: F) -> Result<(DBSPHandle, T), DBSPError>
+    where
+        F: FnOnce(&mut Circuit<()>) -> T + Clone + Send + 'static,
+        T: Clone + Send + 'static,
+    {
+        // When a worker finishes building the circuit, it sends completion status back
+        // to us via this channel.  The function returns after receiving a
+        // notification from each worker.
+        let (init_senders, init_receivers): (Vec<_>, Vec<_>) =
+            (0..nworkers).map(|_| bounded(0)).unzip();
```

Comment:
It's not (yet) clear to me why we'd create a channel per worker, rather than a single channel (which is multi-producer) so all workers can communicate. The effect of this is seen below when receiving the initial  status.

Reply:
A channel per worker means that the client thread will get an error when trying to receive a message from a worker that panic'ed and have a chance to terminate cleanly.  Otherwise, the client would need to receive N messages, one per worker and would deadlock waiting for messages from failed workers.